transaction-script/src/main/java/com/ashishtrivedi16/transactionscript/HotelDaoImpl.java

```diff
@@ -0,0 +1,169 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.ashishtrivedi16.transactionscript;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import javax.sql.DataSource;
+
+public class HotelDaoImpl implements HotelDao {
+
+  private final DataSource dataSource;
+
+  public HotelDaoImpl(DataSource dataSource) {
+    this.dataSource = dataSource;
+  }
+
+  @Override
+  public Stream<Room> getAll() throws Exception {
+    try {
+      var connection = getConnection();
+      var statement = connection.prepareStatement("SELECT * FROM ROOMS");
+      var resultSet = statement.executeQuery(); // NOSONAR
+      return StreamSupport.stream(new Spliterators.AbstractSpliterator<Room>(Long.MAX_VALUE,
+          Spliterator.ORDERED) {
+
+        @Override
+        public boolean tryAdvance(Consumer<? super Room> action) {
+          try {
+            if (!resultSet.next()) {
+              return false;
+            }
+            action.accept(createRoom(resultSet));
+            return true;
+          } catch (Exception e) {
+            throw new RuntimeException(e); // NOSONAR
+          }
+        }
+      }, false).onClose(() -> {
+        try {
+          mutedClose(connection, statement, resultSet);
+        } catch (Exception e) {
+          e.printStackTrace();
+        }
+      });
+    } catch (Exception e) {
+      throw new SqlException(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  public Optional<Room> getById(int id) throws Exception {
+    ResultSet resultSet = null;
+
+    try (var connection = getConnection();
+         var statement = connection.prepareStatement("SELECT * FROM ROOMS WHERE ID = ?")) {
+
+      statement.setInt(1, id);
+      resultSet = statement.executeQuery();
+      if (resultSet.next()) {
+        return Optional.of(createRoom(resultSet));
+      } else {
+        return Optional.empty();
+      }
+    } catch (Exception ex) {
+      throw new SqlException(ex.getMessage(), ex);
+    } finally {
+      if (resultSet != null) {
+        resultSet.close();
+      }
+    }
+  }
+
+  @Override
+  public Boolean add(Room room) throws Exception {
+    if (getById(room.getId()).isPresent()) {
+      return false;
+    }
+
+    try (var connection = getConnection();
+         var statement = connection.prepareStatement("INSERT INTO ROOMS VALUES (?,?,?,?)")) {
+      statement.setInt(1, room.getId());
+      statement.setString(2, room.getRoomType());
+      statement.setInt(3, room.getPrice());
+      statement.setBoolean(4, room.isBooked());
+      statement.execute();
+      return true;
+    } catch (Exception ex) {
+      throw new SqlException(ex.getMessage(), ex);
```

Comment:
Looks a bit suspicious. Why are we catching any exception and throwing `SqlException`? Is `SqlException` caught somewhere?